### ISSUE TITLE
[Resto] Add premature rejuvenation applications module

### ISSUE THREAD

**Issue by buimichael** [WoWAnalyzer/WoWAnalyzer#652]
Note: stuff that changes rejuvenation duration: Persistance relic, Amanthul's wisdom, Flourish, legendary wrists, deep rooted. Stuff that applies rejuvenation: 4PT19, Soul of the archdruid, tearstone, PotA

The expected value from this module should be the amount of seconds potentially lost from premature applications, also suggestions if it's too high.

A suggestion: 
* Extend the rejuvenation.js class to track every single rejuvenation of all players. 
* Keep an array of objects of all players with the current duration left on their rejuvenation. 
* Whenever a target has a rejuv event (refreshbuff/applybuff) they get a timestamp, and whenever an extension is made, such as flourish, deep rooted or shoulders extend their duration. 
* Need to take pandemic into consideration too.
* Calculate if the new application of rejuvenation is within the pandemic time frame.
* Need to take into consideration on rejuvenation proccs, maybe have a variable that track proccPrematureApplications and castPrematureApplications (you can't really control proccs).
* Suggestion on the object: 
[![Screenshot from Gyazo](https://gyazo.com/3449997fcc9be2a33f14ba92ddec0b3a/raw)](https://gyazo.com/3449997fcc9be2a33f14ba92ddec0b3a)

Discuss below if there's something unclear.

**Comment by buimichael** [comment]
Relates to #647 
Some comments from bastas (This is more tied to amanthul's wisdom) but I think it's tied in having this in a rejuvenation tracker.
```
constants:
	- MAX_SHOULDER_PROCS = 3
	- PANDEMIC = 0.3
	- REJUV_BASE_SECONDS_PER_TICK = 3
	- REJUV_BASE_SECONDS = 15;
	- TS_DELTA = 0.33	//at their fastest, rejuvs during bloodlust and EoG will tick roughly 1 time per second. 
						//This adds some wiggle room to make calculations relying on timestamps more accurate, since timestamps can be off by a fraction of a second.
	
	
	
Store the following information about each rejuvenation on each player:
	- ts_startTime		//timestamp when rejuvenation is applied
	- ts_expectedEnd 	//timestamp when rejuvenation would have fallen off without shoulders
	- time_shouldersAdd //how many seconds were added by shoulders. This is used to calculate ts_expectedEnd when shoulders are refreshed during pandemic.
	- numShoulderProcs 	//starts at 0, tracks how many times shoulders have proc'd. Used to calculate ts_shouldersAdd

	
	
Useful functions:
	GetPartialTick(ts_startTime) {

		return partialTick;
	}
	
	//reset the rejuv info for current player :
	reapplyRejuv(player, spellId, ts_expectedEnd) {
		rejuvs[player][spellId].ts_startTime = ts_startTime;
		rejuvs[player][spellId].ts_expectedEnd = ts_expectedEnd + TS_DELTA;
		rejuvs[player][spellId].numShoulderProcs = 0;
		rejuvs[player][spellId].time_shouldersAdd = 0.0;
	}
	
	//get time remaining on a particular rejuv
	GetTimeRemaining( r ) {
		if ( r doesnt exist ) 
			return 0;
		else
			return max(0, r.ts_expectedEnd + r.time_shouldersAdd - ts_current - TS_DELTA);
	}


	
On rejuvenation heal:
	//query for rejuv information, if it doesnt exist there is an error
	r = rejuvs[targetPlayer][spellId];

	//Add shoulder contribution
	if ( r.ts_expectedEnd < ts_current ) {
		shoulderContribution = shoulderContribution + heal.effectiveHealing;
	}
	
	//Update number shoulder procs and calculate ts_shouldersAdd in case this rejuv is prematurely refreshed within pandemic.
	//This is the same as a Warcraftlogs query i use to check number of shoulder procs. I use >=99% to fix rounding error when using hpPercent=100%.
	if ( r.numShoulderProcs < MAX_SHOULDER_PROCS && heal.effectiveHealing == 0 && target.hpPercent >= 99% ) {
		r.numShoulderProcs++;
		
		//Check how many seconds were added by shoulders. This is used in calculating the duration of rejuvs that were refreshed during pandemic.
		partialTick = (ts_current - r.ts_startTime) * ( 1 + hastePercent ) / REJUV_BASE_SECONDS_PER_TICK;
		partialTick = partialTick - floor(partialTick); 
		r.ts_shouldersAdd += (round(1+hastePercent+partialTick)-partialTick) * REJUV_BASE_SECONDS_PER_TICK / (1+hastePercent); //seconds added by shoulder proc
	}
	
	//check for deep rooted
	else if ( target.hpPercent < 35% ) {	
		//calculate new time remaining, this uses similar formula as shoulders
		time_remaining = GetTimeRemaining(r)
		duration = min( (1+PANDEMIC) * (REJUV_BASE_SECONDS+PersistenceLevel), time_remaining + REJUV_BASE_SECONDS + PersistenceLevel );
		ticks = duration*(1+hastePercent) / REJUV_BASE_SECONDS_PER_TICK;
		roundedDuration = round(ticks)*REJUV_BASE_SECONDS_PER_TICK / (1+hastePercent);
		
		//update rejuv values
		ts_nextExpectedEnd = ts_current + roundedDuration;
		reapplyRejuv(targetPlayer, spellId, ts_nextExpectedEnd); //reset rejuv information
	}
	
On rejuvenation refreshed or freshly applied: 
	//query for rejuv info
	r = rejuvs[targetPlayer][spellId];
	
	//reapply rejuvenation
	time_remaining = GetTimeRemaining(r);
	newDuration = min( (1+PANDEMIC) * (REJUV_BASE_SECONDS+PersistenceLevel), time_remaining + REJUV_BASE_SECONDS + PersistenceLevel );
	ts_newExpectedEnd = ts_current + newDuration;
	reapplyRejuv(player, spellId, ts_newExpectedEnd); //reset rejuv information

	
On flourish cast:
	foreach ( rejuvenation r in rejuvs ) //add 6 seconds and round it. {
		newDuration = r.ts_expectedEnd + ts_shouldersAdd + 6 - r.ts_startTime - TS_DELTA;
		ticks = newDuration*(1+hastePercent) / REJUV_BASE_SECONDS_PER_TICK;
		roundedDuration = round(ticks)*REJUV_BASE_SECONDS_PER_TICK / ( 1 + hastePercent );
		
		//update expected end, but dont reset rejuv info entirely.
		r.ts_expectedEnd = r.ts_startTime + roundedDuration + TS_DELTA;
	}
```
very possibly a few bugs in there, but the idea overall will work.  This assumes you can query current timestamp of a healing event, a player's haste % at a timestamp (this can be calculated the hard way if this isnt possible), and a player's persistence level.
ideas:
- Track expectedExpiration, numberOfShoulderProcs, and SecondsAddedByShoulders on every rejuvenation 
- Update expectedExpiration on flourish, deep rooted, or rejuvenation reapplications/refreshes
- If a rejuv does healing after its expected expiration, this is appilcable to shoulders.
- Flourish, Deep rooted, and shoulder procs all round away partial ticks, which requires knowledge of haste%
- Because timestamps can be inaccurate by a fraction of a second, we add a small delta amount (0.33) to expectedExpiration when checking if current tick is after expectedExpiration. Since rejuvs tick infrequently (>1s between ticks), this will never be an issue, and will ensure we dont overcalculate shoulder amount accidentally.

**Comment by kfinch** [comment]
`Flourish, Deep rooted, and shoulder procs all round away partial ticks, which requires knowledge of haste%` - can you explain what you mean by this?

**Comment by tremaho** [comment]
The way these HoT extensions work in-game is by adding _t_ seconds and then rounding to the nearest whole number of ticks. Current haste percentage is used for this rounding (see below). We could just say flourish is always +6 seconds, or that a shoulder extension is always +3 seconds, but it wouldn't always be correct without reproducing the rounding algorithm. Deep rooted also does this rounding after adding +_baseduration_ clamped by pandemic _1.3*baseduration_.

An example of shoulders rounding:
- Let **R** =5.4s (time remaining on current rejuvenation)
- Let **H** = 20% (current haste percentage)
- Now let the shoulders proc an extension:
    - `(R+3.0s) = 8.4` is the time left before rounding
    - `(8.4)*(1+H)/3 = (8.4)*(1.2)/3 = 3.36` is the number of ticks left at 8.4 seconds with **H**=20% haste
    - `Round(3.36) = 3.0` is the number of ticks left after rounding
    - `3.0*3/1.2 = 7.5` is the time left after rounding
    - `7.5 - 5.4 = 2.1` is the time added by this shoulder extension
- So we know to attribute the last 2.1 seconds of this rejuvenation's healing to Aman'Thul's Wisdom.
- An estimation without doing this rounding would attribute the last 3.0 seconds of this rejuvenation to Aman'Thul's Wisdom. 
- Let's compare the two approaches assuming haste remains constant:
    - If we capture the last 3.0s, we will capture the last 2 ticks of this rejuvenation.
    - If we capture the last 2.1s, we will capture just the last tick of this rejuvenation
    - In other words, by not rounding we overestimate Aman'Thul's Wisdom value.
- If our haste changes after the rounding, we still want to attribute the last 2.1s to Aman'Thul's Wisdom. This could mean capturing more or less ticks.

**Comment by kfinch** [comment]
@tremaho Thanks for the detailed explanation! I knew that Shoulders extensions did something silly with haste, but I had no idea it applied to Flourish and Deep Rooted as well. As I plan to soon be implementing this, I have a few followup questions / comments

- You say it rounds, but just to be absolutely clear that means it can round up as well as down, right?
- How do partial ticks work? My understanding was that if a HoT ticks with less than a full tick of time remaining afterwards, it will do a 'partial' tick right before expiring. Like if player has no haste and so 3 second rejuv ticks, and rejuv ticks with 1.5 seconds left, it would do a half strength tick right before expiring.
- Do partial ticks hold up as I think in the edge cases? For example if a HoT has 1.01 ticks of time left, will it really do a normal tick and then almost immediately do a tiny tick before falling?
- Would capturing the last N seconds really be accurate due to the granularity of ticks, / partial ticks or are we assuming it will smooth out over the sample size of many ticks? I feel like naively attributing healing that happens in the given window might introduce systemic error.
- The solution I was thinking about for capturing healing attributable to in the last N seconds was to capture healing done during the period and also the last tick before the period, then scaling appropriately. For example, our heal ticks at .... 3.5, 2.5, 1.5, 0.5, and (partial) 0.0 seconds before expiration, and we're capturing the healing attributable to the last 2.0 seconds. I would sum the healing of the 2.5, 1.5, 0.5, and 0.0 ticks, and then multiply by (2.0 / 2.5). Let me know what you think of this method.

**Comment by tremaho** [comment]
- Can it round both ways? Yes `Round(8.49) = 8.00,  Round(8.50) = 9.00`
- Even a partial tick of < 0.01 appears to happen separately in the combat log, even though they heal almost instantaneously:
![partialtick](https://user-images.githubusercontent.com/32786336/32755065-c3d1102c-c890-11e7-9a0b-40b73a294c90.png)
- Rounding removes all partial ticks, but changes in haste can reintroduce them.
- Is tracking healing over a given window inaccurate and/or could it produce systemic error? 
  - Unfortunately, yes. At 0% haste your rejuvenation doesnt tick at exactly 3 second intervals. 
  - There is minor variance based on when the server processes the heal, which is why you have to do some fuzzy logic by adding a small amount to both ends of the interval you want to check. 
  - One reason this is "ok" is that rejuvenation will tick infrequently relative to the small amount you have to add to the interval.
  - Technically, the Tearstone module does something similar by capturing rejuvenations within a small 200ms window of wild growth.

Figuring out the best implementation is important, and to be honest, I'm really not sure if my approach is the best. With my method, I was just trying to think of a way to answer the statistical question:  
- If I didn't equip shoulders and cast the same spells at the same times, how much less healing would I have done?

It's possible that answering a different question is more useful and/or easier to implement.

**Comment by kfinch** [comment]
Hey, an update on this. I have a HoT attributor module implemented and PR is up here: #1161 

Extension implementing part isn't yet implemented, but you can see the TODOs. Hope to get this finished sooner rather than later.
